client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -285,45 +302,52 @@ public HonoClient getOrCreateEventSender(
     void getOrCreateSender(final String key, final Consumer<Handler<AsyncResult<MessageSender>>> newSenderSupplier,
             final Handler<AsyncResult<MessageSender>> resultHandler) {
 
-        final MessageSender sender = activeSenders.get(key);
-        if (sender != null && sender.isOpen()) {
-            LOG.debug("reusing existing message sender [target: {}, credit: {}]", key, sender.getCredit());
-            resultHandler.handle(Future.succeededFuture(sender));
-        } else if (!isConnected()) {
-            resultHandler.handle(Future.failedFuture(new ServerErrorException(
-                    HttpURLConnection.HTTP_UNAVAILABLE, "no connection to service")));
-        } else if (!creationLocks.computeIfAbsent(key, k -> Boolean.FALSE)) {
-
-            // register a handler to be notified if the underlying connection to the server fails
-            // so that we can fail the result handler passed in
-            final Handler<Void> connectionFailureHandler = connectionLost -> {
-                // remove lock so that next attempt to open a sender doesn't fail
-                creationLocks.remove(key);
-                resultHandler.handle(Future.failedFuture(
-                        new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "no connection to service")));
-            };
-            creationRequests.add(connectionFailureHandler);
-            creationLocks.put(key, Boolean.TRUE);
-            LOG.debug("creating new message sender for {}", key);
-
-            newSenderSupplier.accept(creationAttempt -> {
-                if (creationAttempt.succeeded()) {
-                    MessageSender newSender = creationAttempt.result();
-                    LOG.debug("successfully created new message sender for {}", key);
-                    activeSenders.put(key, newSender);
-                } else {
-                    LOG.debug("failed to create new message sender for {}", key, creationAttempt.cause());
-                    activeSenders.remove(key);
-                }
-                creationLocks.remove(key);
-                creationRequests.remove(connectionFailureHandler);
-                resultHandler.handle(creationAttempt);
-            });
+        synchronized (connectionLock) {
+            if (!isConnected()) {
+                resultHandler.handle(Future.failedFuture(new ServerErrorException(
+                        HttpURLConnection.HTTP_UNAVAILABLE, "no connection to service")));
+            } else if (context == null) {
+                throw new IllegalStateException("vert.x context is null despite existing connection");
+            } else {
+                context.runOnContext(get -> {
+                    final MessageSender sender = activeSenders.get(key);
```

Comment:
Wouldn't it be possible at this location that `isConnected()` has changed. And the following access to internal structured would be un-synchronized as the context might be a different one, it can be set with `setConnection`. So this could run on different threads.